openstack-api-charm-creation-guide.md

```diff
@@ -0,0 +1,240 @@
+
+```
+class CongressCharmFactory(OpenStackCharmFactory):
+
+    releases = {
+        'liberty': CongressCharm
+    }
+
+    first_release = 'liberty'
+```
+
+## Add Congress code to react to events
+
+### Install Congress Packages
+
+The reactive framework is going to emit events that the Congress charm can react to. the charm needs to define how its going to react to these events and also raise new events as needed.
+
+The first action a charm needs to do is to install the Congress code. This is by done running the install method from CongressCharm created earlier. Edit charm/reactive/handlers.py ...
+
+```
+import charms.reactive as reactive
+import charm.openstack.congress as congress
+import charmhelpers.contrib.openstack.utils as ch_utils
+
+charm = None
+
+def get_charm():
+    # Retrieve a charm instance from the charm factory.
+    global charm
+    if charm is None:
+        charm = congress.CongressCharmFactory.charm(
+            release=ch_utils.os_release('congress-common', base='liberty'))
+    return charm
+
+@reactive.hook('install')
+def install_packages():
+    # When install hook fires install charm packages
+    get_charm().install()
+```
+
+### Configure Congress Relation
+
+At this point the charm could be built and deployed and it would deploy a unit, and install congress. However there is no code to specify how this charm should interact with the services to depend on. For example when joining the database the charm needs to specify the user and database it requires. The following code configures the relations with the dependant services.
+
+```
+@reactive.when('amqp.connected')
+def setup_amqp_req(amqp):
+    # Request username and vhost from rabbit charm
+    amqp.request_access(username='congress',
+                        vhost='openstack')
+
+@reactive.when('shared-db.connected')
+def setup_database(database):
+    # Request username and database from database charm
+    database.configure('congress', 'congress', hookenv.unit_private_ip())
+
+@reactive.when('identity-service.connected')
+def setup_endpoint(keystone):
+    # Regester endpoints with keystone and receive service credentials
+    charm = get_charm()
+    keystone.register_endpoints(charm.service_type,
+                                charm.region,
+                                charm.public_url,
+                                charm.internal_url,
+                                charm.admin_url)
+```
+
+## Configure Congress
+
+Now that the charm has the relations defined that it needs the Congress charm is in a postion to generate its configuration files.
+
+### Create templates
+
+The charm code searches through the templates directories looking for a directory cooresponding to the Openstack release being installed
+or earlier. Since Liberty is the earliest release the charm is supporting a directory called liberty will house the templates.
+
+```mkdir -p templates/liberty
```

Comment:
Maybe add an apt-install congress-server (on xenial) here to get the config files installed that are being copied to templates/mitaka.